File: 이것이 취업을 위한 코딩 테스트다/BFS_DFS/미로_탈출/미로_탈출.cpp
#include <iostream>
#include <queue>
using namespace std;

int rowSize, columnSize;
int graph[201][201];

void bfs(int newRow, int newColumn) {
  int row[] = {-1, 1, 0, 0};
  int column[] = {0, 0, -1, 1};
  
  queue<pair<int, int>> q;
  q.push({newRow, newColumn});

  while(!q.empty()) {
    int currentRow = q.front().first;
    int currentColumn = q.front().second;
    q.pop();
    for (int i = 0; i < 4; i++) {
      int nextRow = currentRow + row[i];
      int nextColumn = currentColumn + column[i];

      if (-1 >= nextRow || nextRow >= rowSize || -1 >= nextColumn || nextColumn >= columnSize) {
        continue;
      }

      if (graph[nextRow][nextColumn] == 1) {
        graph[nextRow][nextColumn] = graph[currentRow][currentColumn] + 1;
        q.push({nextRow, nextColumn});
      }
    }
  }
}

int main() {
  cin >> rowSize >> columnSize;

  for (int row = 0; row < rowSize; row++) {
    for (int column = 0; column < columnSize; column++) {
      scanf("%1d", &graph[row][column]);
    }
  }

  bfs(0, 0);

  cout << graph[rowSize -1][columnSize - 1];
}

File: 이것이 취업을 위한 코딩 테스트다/BFS_DFS/미로_탈출/미로_탈출.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class 미로_탈출 {
  public static int rowSize, columnSize;
  public static int[][] graph = new int[200][200];
  public static void bfs(int newRow, int newColumn) {
    int []row = {-1, 1, 0, 0};
    int []column = {0, 0, -1, 1};

    Queue<ArrayList<Integer>> q = new LinkedList<ArrayList<Integer>>();
    q.offer(new ArrayList<Integer>(Arrays.asList(newRow, newColumn)));

    while(!q.isEmpty()) {
      int currentRow = q.peek().get(0);
      int currentColumn = q.peek().get(1);
      q.poll();

      for(int i = 0; i < 4; i++) {
        int nextRow = currentRow + row[i];
        int nextColumn = currentColumn + column[i];

        if(-1 >= nextRow || nextRow >= rowSize || -1 >= nextColumn || nextColumn >= columnSize) {
          continue;
        }

        if(graph[nextRow][nextColumn] == 1) {
          graph[nextRow][nextColumn] = graph[currentRow][currentColumn] + 1;
          q.offer(new ArrayList<Integer>(Arrays.asList(nextRow, nextColumn)));
        }
      }
    }
  }
  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);

    rowSize = sc.nextInt();
    columnSize = sc.nextInt();
    sc.nextLine();

    for(int row = 0; row < rowSize; row++) {
      String str = sc.nextLine();

      for(int column = 0; column < columnSize; column++) {
        graph[row][column] = str.charAt(column) - '0';
      }
    }

    bfs(0, 0);

    System.out.println(graph[rowSize-1][columnSize-1]);
  }
}
